core/trino-main/src/main/java/io/trino/sql/planner/PlanFragmenter.java

```diff
@@ -370,7 +372,13 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
                     .map(PlanFragment::getId)
                     .collect(toImmutableList());
 
-            return new RemoteSourceNode(exchange.getId(), childrenIds, exchange.getOutputSymbols(), exchange.getOrderingScheme(), exchange.getType());
+            return new RemoteSourceNode(
+                    exchange.getId(),
+                    childrenIds,
+                    exchange.getOutputSymbols(),
+                    exchange.getOrderingScheme(),
+                    exchange.getType(),
+                    context.get().getPartitioningHandle().isCoordinatorOnly() ? getRetryPolicy(session) : RetryPolicy.NONE);
```

Comment:
Is `context.get().getPartitioningHandle().isCoordinatorOnly()` another way of saying "is the the task the root of the execution, sending data to the user"?

If so isn't it exploitation of the abstration. Is it prohibited for plan to have "coordinator-only" stages which are not root of execution.

Reply:
Agreed. The condition should be "If this is a worker to coordinator boundary - add a deduplication buffer". Let me change that